proposals/0002.md

```diff
@@ -0,0 +1,272 @@
+
+`source` should return a value when the error has another error nested inside.
+The wrapped error should be the cause of the outer one, usually a bit lower level and more detailed.
+The errors can be wrapped multiple times to express the whole chain of errors down to the root
+cause. It can be traversed by calling `source` recursively.
+
+For example the custom `FileLoadError` can wrap the `std::io::Error`.
+This way the file loading function doesn't have to know about all the possible file system failures
+described by the `std::io::Error`.
+When it gets one, it simply wraps it in the `FileLoadError` and returns it.
+If a higher level function is interested, what caused the `FileLoadError`, it can call its `source`
+method and get its cause.
+
+`source` should return `None` when the error is the root cause and doesn't wrap another error.
+It's also impossible to return a wrapped error when it doesn't implement `Error`.
+
+### The bevavior of the `Display`
+
+The `Display` implementation should briefly describe the error.
+It should never include information provided by the error returned from the `source`.
+Its output should be formatted as a single line, preferably a single sentence.
+If it's a single sentence, it should start with a lowercase letter and end with no interpunction.
+
+Examples:
+
+`failed to load the file '/tmp/file'` **good**
+
+`failed to load the file '/tmp/file': file not found`
+**bad**: includes information provided by its `source`
+
+`Failed to load the file '/tmp/file'!`
+**bad**: a single sentence starting with a capital letter and ending with interpunction
+
+`Failed to load the file. Path: '/tmp/file'.`
+**ugly**: it has two sentences
+
+## Printing a chain of `Error`s
+
+When all the errors in a chain behave correctly, insightful and well formatted messages can be
+easily provided to the user.
+The error can be iterated by recursively calling `source` to print each layer and provide the full
+chain of errors that caused the failure.
+Due to strict `Display` implementation rules there's a lot of creative freedom when it comes to
+chain formatting.
+The `source` traverasal can be done manually, with `anyhow::Chain` or with very similar
+`std::error::Chain` as soon as it's stabilized.
+
+An example of the manual chain printing:
+
+```rust
+println!("An error occured: {}", error);
+let mut parent = &error;
+while let Some(cause) = parent.source() {
+    println!("  caused by: {}", cause);
+    parent = cause;
+}
+```
+
+An example output:
+
+```
+An error occured: failed to load the configuration file
+  caused by: failed to load the file '~/.app_config'
+  caused by: file not found
+```
+
+An example of the chain printing with `anyhow::Chain`:
+
+```rust
+print!("An error occured: {}", error);
+anyhow::Chain::new(&error).for_each(|e| print!(": {}", e);
+println!();
+```
+
+An example output:
+
+```
+An error occured: failed to load the configuration file: failed to load the file '~/.app_config': file not found
```

Comment:
Wait, why would we not use `println!` here?

Reply:
Because then we would get a multiline output. I wanted to show that the full error chain can be printed as a single line too, e.g. for a logger.

Reply:
Ah ok. I'd suggest not mixing examples. I saw this as an example for the `anyhow::Chain` capability. I would maybe add another one before this comparing `anyhow::Chain` to the `Error::Chain` example. Then you show this one with a single line.

Reply:
As another note here: I find the `:` punctuation here ugly on the single line :sweat_smile:

Reply:
Added the third example for clarity. Switched `:` to `⮩` for that aesthetic feel. Is it good now?